apps/src/lib/node/ledger/shell/vote_extensions/bridge_pool_vext.rs

```diff
@@ -0,0 +1,602 @@
+        });
+        current_validators.data.insert(
+            1,
+            Some(pos::types::ValidatorSet {
+                active: vals,
+                inactive: Default::default(),
+            }),
+        );
+        shell.storage.write_validator_set(&current_validators);
+
+        // register Bertha's protocol key
+        let pk_key = protocol_pk_key(&bertha_address());
+        shell
+            .storage
+            .write(
+                &pk_key,
+                bertha_keypair()
+                    .ref_to()
+                    .try_to_vec()
+                    .expect("Test failed."),
+            )
+            .expect("Test failed.");
+
+        // change pipeline length to 1
+        let mut params = shell.storage.read_pos_params();
+        params.pipeline_len = 1;
+
+        // register Bertha's consensus key
+        let consensus_key = gen_keypair();
+        shell.storage.write_validator_consensus_key(
+            &bertha_address(),
+            &ValidatorConsensusKeys::init(consensus_key.ref_to(), 0, &params),
+        );
+
+        // register Bertha's ethereum keys.
+        let hot_key = gen_secp256k1_keypair();
+        let cold_key = gen_secp256k1_keypair();
+        shell.storage.write_validator_eth_hot_key(
+            &bertha_address(),
+            &ValidatorEthKey::init(hot_key.ref_to(), 0, &params),
+        );
+        shell.storage.write_validator_eth_cold_key(
+            &bertha_address(),
+            &ValidatorEthKey::init(cold_key.ref_to(), 0, &params),
+        );
+
+        // we advance forward to the next epoch
+        let mut req = FinalizeBlock::default();
+        req.header.time = namada::types::time::DateTimeUtc::now();
+        shell.storage.last_height = BlockHeight(15);
+        shell.finalize_block(req).expect("Test failed");
+        shell.commit();
+        assert_eq!(shell.storage.get_current_epoch().0.0, 1);
+
+        // Check that Bertha's vote extensions pass validation.
+        let to_sign = [[0; 32], Uint::from(0).to_bytes()].concat();
+        let sig =
+            Signed::<Vec<u8>, SignableEthBytes>::new(&hot_key, to_sign).sig;
+        let vote_ext = bridge_pool_roots::Vext {
+            block_height: shell.storage.get_current_decision_height(),
+            validator_addr: bertha_address(),
+            sig,
+        }
+        .sign(&bertha_keypair());
+
+        assert!(shell.validate_bp_roots_vext(
+            vote_ext,
+            shell.storage.get_current_decision_height(),
+        ));
+    }
+
+    /// Test that the function crafting the bridge pool root
+    /// vext creates the expected payload. Check that this
+    /// payload passes validation.
+    #[test]
+    fn test_happy_flow() {
+        let (shell, _, _) = setup_at_height(3u64);
+        let address = shell
+            .mode
+            .get_validator_address()
+            .expect("Test failed")
+            .clone();
+        let to_sign = [[0; 32], Uint::from(0).to_bytes()].concat();
```

Comment:
I've seen this line a couple of times. it could be a separate func, that returns the `to_sign` content, in case we change this in the future, for some reason. it would make refactoring easier, then